### PR TITLE
Support for HTTP HEAD method, ETags, and If-None-Match conditional requests

### PR DESCRIPTION
**Pull request by veloman-yunkan**
Will fix kiwix/kiwix-tools#140

This is the last batch of commits from #340 following #344 and #345.

It is easier to review this PR one commit at a time.

Implementation of the support for the HEAD request method is far from optimal. It executes all of the code of the GET request method and then discards the body. This optimizes the traffic but not the server CPU usage. However, if clients properly take advantage of ETags set by the server they should NOT abuse HEAD requests, but issue conditional GET requests with the If-None-Match header set to the value of the ETag from the response to the original GET request.

Implementation-wise the ETag string used by Kiwix server (more precisely, its value inside the double quotes) consists of two parts:

1. ServerId - A "unique" string obtained on server start up
2. Options -  Zero or more characters encoding the values of some of the headers of the response

The two parts are separated with a slash (/) symbol (which is always present, even when the the options part is empty). Neither portion of a Kiwix ETag may contain the slash symbol.

Examples of valid Kiwix server ETags (including the double quotes):

  "abcdefghijklmn/"
  "1234567890/z"
  "1234567890/cz"

The options part of the Kiwix ETag allows to correctly set the required headers when responding to a conditional If-None-Match request with a 304 (Not Modified) response without following the full code path that would discover the necessary options.

This PR comes with a set of new unit-tests covering the changes almost fully (several omitted edge cases related to invalid and weak ETags in the If-None-Match header leave a couple of lines uncovered).

### CODE REVIEW

File: src/server/etag.cpp [hunk 1]
```diff
@@ -61,4 +89,47 @@ std::string ETag::get_etag() const
...
+  std::string etag_str;
+  while ( ss >> etag_str )
+  {
+    if ( etag_str.back() == ',' )
+      etag_str.pop_back();
+
+    const ETag etag = parse(etag_str);
+    if ( etag && etag.m_serverId == server_id )
+      return etag;
+  }
```
Comment (mgautierfr) [inline src/server/etag.cpp:130]: Should we check the options also ?

File: src/server/response.cpp [hunk 2]
```diff
@@ -196,6 +200,7 @@ Response::create_raw_content_mhd_response(const RequestContext& request)
          It has no incidence on other browsers
          See http://www.subbu.org/blog/2008/03/ie7-deflate-or-not and comments */
       m_content = string((char*)&compr_buffer[2], comprLen - 2);
+      m_etag.set_option(ETag::COMPRESSED_CONTENT);
     } else {
       shouldCompress = false;
```
[the remark this answers is not in the document]
Comment (veloman-yunkan) [reply src/server/response.cpp:205]: The `if` of this `else` reads

```
if (err == Z_OK && comprLen > 2 && comprLen < (m_content.size() + 2)) {
```

Thus the `else` branch corresponds to the case when an attempt to compress the content didn't reduce its size, in which compression is not justified and should be abandoned.

File: src/server/response.cpp [hunk 3]
```diff
@@ -204,9 +209,11 @@ Response::create_raw_content_mhd_response(const RequestContext& request)
   MHD_Response* response = MHD_create_response_from_buffer(
     m_content.size(), const_cast<char*>(m_content.data()), MHD_RESPMEM_MUST_COPY);
 
-  if (shouldCompress) {
+  if ( m_etag.get_option(ETag::COMPRESSED_CONTENT) ) {
```
[the remark this answers is not in the document]
Comment (veloman-yunkan) [reply src/server/response.cpp:212]: Yes, if `m_etag` was set while handling a conditional If-None-Match request and the compression option was obtained from the ETag value. In this case the response is a 304 (Not Modified) one with empty body which shouldn't be compressed.

File: src/server/etag.cpp (hunk 1 above)
Comment (veloman-yunkan) [reply src/server/etag.cpp:130]: No. There are no known options at this point. The server has to pick from the list of ETags coming from the If-None-Match header the one (if any) that was generated by this server instance and use its options part for producing a correct 304 (Not Modified) response by correctly setting all the required headers.

File: src/server/etag.cpp [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+
+namespace {
+
+// Characters in the options part of the ETag could in principle be picked up
+// from the latin alphabet in natural order (the character corresponding to
+// ETag::Option opt would be 'a'+opt; that would somewhat simplify the code in
+// this file). However it is better to have some mnemonics in the option names,
+// hence below variable: all_options[opt] corresponds to the character going
+// into the ETag for ETag::Option opt.
+const char all_options[] = "cz";
```
[the remark this answers is not in the document]
Comment (veloman-yunkan) [reply src/server/etag.cpp:40]: I had (almost) such a version (see https://github.com/kiwix/kiwix-lib/commit/b56495eb952b706386900c65f26bcb868ddc1dc8#diff-5abb299a97e453d2845d6d292fee4f6dL50-R60). I switched to the current implementation with the purpose of implementing the [isValidOptionsString()](https://github.com/kiwix/kiwix-lib/pull/348/files#diff-403f9fd39b2753cc313043eddf4cf687R62-R65) function in such a way that one doesn't have to change it if more options must be added in the future. Then adding an option will require only introducing a new enum member and updating `all_options` (the latter part will be automatically reminded by the compiler due to the [`static_assert()` below the definition of that string](https://github.com/kiwix/kiwix-lib/pull/348/files#diff-403f9fd39b2753cc313043eddf4cf687R42)).

File: src/server/etag.h [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+//
+// The options part of the Kiwix ETag allows to correctly set the required
+// headers when responding to a conditional If-None-Match request with a 304
+// (Not Modified) response without following the full code path that would
+// discover the necessary options.
+
+class ETag
+{
+  public: // types
+    enum Option {
```
[the remark this answers is not in the document]
Comment (veloman-yunkan) [reply src/server/etag.h:53]: Then I will have to write `ETag::Option::CACHEABLE_ENTITY` instead of `ETag::CACHEABLE_ENTITY` which IMHO is less readable. `ETag` is a small class and serves as a scope for its (only) enum member type. This can be done if another `enum` type must be introduced in `ETag`.

File: src/server/response.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (veloman-yunkan) [reply src/server/response.cpp:218]: I added a comment clarifying that point in the code

File: src/server/response.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (mgautierfr) [reply src/server/response.cpp:205]: Ho, I miss this with the diff view not displaying all the code.

**Comment by mgautierfr** [comment]
Seems good to me. You answer all my questions.
